backend/src/lib.rs

```diff
@@ -1,104 +1,176 @@
-impl ProverWithoutParams for Halo2MockBackend {
-    fn prove<T: FieldElement>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-    ) -> Option<Proof> {
-        halo2::mock_prove(pil, fixed, witness);
-        None
+/// Concrete dynamic dispatch Backend object, for backends with setup.
+struct ConcreteBackendWithSetup<B>(B);
+
+/// Concrete implementation for backends with setup.
+impl<F: FieldElement, B: BackendWithSetup<F>> Backend<F> for ConcreteBackendWithSetup<B> {
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()> {
+        self.0.prove(pil, fixed, witness, prev_proof, output_dir)
     }
-}
 
-#[cfg(feature = "halo2")]
-impl ProverAggregationWithParams for Halo2AggregationBackend {
-    fn prove<T: FieldElement, R1: io::Read, R2: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        proof: R1,
-        params: R2,
-    ) -> Proof {
-        halo2::prove_aggr_read_proof_params(pil, fixed, witness, proof, params)
+    fn write_setup(&self, output: &mut dyn io::Write) -> Result<(), Error> {
+        Ok(self.0.write_setup(output)?)
     }
 }
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("input/output error")]
+    IO(#[from] std::io::Error),
+    #[error("the backend has not setup operations")]
+    NoSetupAvailable,
+}
+
+pub type Proof = Vec<u8>;
+
+/// Dynamic interface for a backend.
+pub trait Backend<F: FieldElement> {
+    /// Perform the backend proving. If prev_proof is provided, proof
+    /// aggregation is performed.
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()>;
+
+    /// Write the prover setup to a file, so that it can be loaded later.
+    fn write_setup(&self, output: &mut dyn io::Write) -> Result<(), Error>;
+}
+
+/// Dynamic interface for a backend factory.
+pub trait BackendFactory<F: FieldElement> {
+    /// Maybe perform the setup, and create a new backend object.
+    fn new(&self, degree: DegreeType) -> Box<dyn Backend<F>>;
+
+    /// Create a backend object from a prover setup loaded from a file.
+    fn new_from_setup(&self, input: &mut dyn io::Read) -> Result<Box<dyn Backend<F>>, Error>;
+}
+
+/// Trait implemented by all backends.
+trait BackendImpl<F: FieldElement> {
+    /// Perform the setup and create a new backend object.
+    fn new(degree: DegreeType) -> Self;
+
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()>;
```

Comment:
why doesn't it return a proof?

Reply:
See this: https://github.com/powdr-labs/powdr/pull/417#discussion_r1284662092